[lldb-dap] Add an option to provide a format for stack frames

walter-erquinigo opened the pull request:
When this option gets enabled, descriptions of stack frames will be generated using the format provided in the launch configuration instead of simply calling `SBFrame::GetDisplayFunctionName`. This allows lldb-dap to show an output similar to the one in the CLI.

Commit:
[lldb-dap] Add an option to show function args in stack frames

When this option is enabled, display names of stack frames are generated using the `${function.name-with-args}` formatter instead of simply calling `SBFrame::GetDisplayFunctionName`. This makes lldb-dap show an output similar to the one in the CLI.

This option is disabled by default because of its performance cost. It's a good option for non-gigantic programs.

clayborg reviewed (changes requested):
So this hardcodes the frame display to only be the function name, or function name with args. Do we want to allow users to specify a frame format in the lldb-vscode settings? We could have two settings: one to enable the feature with something like `showFramesWithCustomFormat` and on that contains the format string like `customFrameFormat` which would default to `"${function.name-with-args}"`. Then users have complete control over how and what gets displayed. Some people might want to show the pc value and then the function name like:
```
  "customFrameFormat": "${frame.pc} ${function.name-with-args}"
```
If we want this then we will want to modify the SBFrame changes to be something like:
```
 SBError SBFrame::ApplyFormat(SBStream &output, const char *format_string);
```
We might even consider makding a `lldb::SBFormat` class that allows us to compile a format string into an object and then use it. If we do this then the API above will be:
```
SBError SBFrame::ApplyFormat(SBStream &output, lldb::SBFormat format);
```

bulbazord commented on lldb/include/lldb/API/SBFrame.h:
```diff
@@ -87,8 +87,15 @@ class LLDB_API SBFrame {
+  /// Similar to \a GetDisplayFunctionName() but with function arguments and
+  /// their values inserted into the function display name whenever possible.
+  ///
+  /// \param[in] output
```
this should be `\param[out]`

bulbazord commented on lldb/include/lldb/API/SBFrame.h:
```diff
@@ -87,8 +87,15 @@ class LLDB_API SBFrame {
+  ///
+  /// \param[in] output
+  ///   The stream where the display name is written to.
+  void GetDisplayFunctionNameWithArgs(SBStream &output);
```
Should this return an `SBError`? How do I know what went wrong if this fails?

bulbazord commented on lldb/include/lldb/Target/StackFrame.h:
```diff
@@ -324,8 +325,23 @@ class StackFrame : public ExecutionContextScope,
+  /// StackFrame::Dump()), in which case \b false is returned. Otherwise, \b
+  /// true is returned.
+  ///
+  /// \param[in] strm
```
This is an out parameter no?

bulbazord commented on lldb/source/API/SBFrame.cpp:
```diff
@@ -1232,6 +1234,32 @@ const char *SBFrame::GetFunctionName() const {
+      if (frame) {
+        FormatEntity::Entry format;
+        Status s = FormatEntity::Parse("${function.name-with-args}", format);
+        assert(
+            s.Success() &&
+            "The ${function.name-with-args} format must be parsed correctly");
```
I don't think this should be an assertion. If something internally went wrong with LLDB, we shouldn't crash. This might be a good place to return an SBError

bulbazord commented on lldb/test/API/tools/lldb-dap/stackTrace/TestDAP_stackTrace.py:
```diff
@@ -187,3 +188,19 @@ def test_stackTrace(self):
             0, len(stackFrames), "verify zero frames with startFrame out of bounds"
         )
+
+    @skipIfWindows
```
Why skip on Windows?

walter-erquinigo replied:
I think all vscode tests skip on windows...

walter-erquinigo commented:
@clayborg , I did pretty much what you asked but with a few changes:

- I added the `customFrameFormat` option to the json config, and it'll be used if provided and non-empty. I just find it simpler to control everything with one single option.
- I added the SBFormat class that wraps around FormatEntity::Entry. I had to convert FormatEntity into a namespace to be able to use forward declarations, but the change was extremely trivial because FormatEntity was just a bag of static functions.
- lldb-dap will dump an error on the debug console if the provided format is wrong, and the original flow is used in this case.

clayborg commented on lldb/include/lldb/API/SBFrame.h:
```diff
@@ -193,6 +193,27 @@ class LLDB_API SBFrame {
+  /// \param[out] output
+  ///   The stream where the description will be written to.
+  ///
+  /// \param[in] default_value
+  ///   If the description couldn't be generated, and this parameter is not
+  ///   null, it will be printed to the \p output stream. This doesn't affect
+  ///   the return value of this method.
```
Do we need this parameter? If we return `false`, then the user can do what ever they need to. Not sure how to provide a good `default_value` for a frame that makes any sense as it can't be hard coded, and we don't want to generate a frame string before calling this just so we can use it as the default_value? I would prefer to remove this paramter and fix any call sites to do some other logging if false is returned

clayborg commented on lldb/include/lldb/Target/StackFrame.h:
```diff
@@ -324,8 +325,29 @@ class StackFrame : public ExecutionContextScope,
+  /// \param[in] frame_marker
+  ///   Optional string that will be prepended to the frame output description.
+  ///
+  /// \param[in] use_fallback_format_on_error
+  ///   If dumping with the given \p format fails and this flag is enabled, then
+  ///   dumping will be retried with a default fallback format.
```
Maybe change the wording so this will use the current `frame-format` from the settings? The term "fallback format" doesn't clearly indicate what is going to happen here. Maybe:
```
  /// \param[in] use_frame_format_from_settings_on_error
  ///   If dumping with the given \p format fails and this flag is enabled, then
  ///   dumping will be retried with the format from the "frame-format" setting.

clayborg commented on lldb/source/API/SBFormat.cpp:
```diff
@@ -0,0 +1,50 @@
+  Status status = FormatEntity::Parse(format, *format_entry_sp);
+
+  SBFormat sb_format;
+  if (status.Fail())
+    error.SetError(status);
+  else
+    sb_format.m_opaque_sp = format_entry_sp;
```
We probably should always set the error in case the incoming `error` is already in an error state, and we should always set or clear the m_opaque_sp if things fail? Or we can leave m_opaque_sp alone. I was thinking something like:
```
error.SetError(status);
if (error.Success())
  sb_format.m_opaque_sp = format_entry_sp;
else
  sb_format.m_opaque_sp.clear();
```

clayborg commented on lldb/source/API/SBFrame.cpp:
```diff
@@ -601,8 +602,8 @@ SBValue SBFrame::FindValue(const char *name, ValueType value_type,
                 &variable_list);
-          if (value_type == eValueTypeVariableGlobal 
-              || value_type == eValueTypeVariableStatic) {
+          if (value_type == eValueTypeVariableGlobal ||
+              value_type == eValueTypeVariableStatic) {
```
revert? not related to this change?

clayborg commented on lldb/source/API/SBFrame.cpp:
```diff
@@ -814,9 +815,11 @@ SBValueList SBFrame::GetVariables(const lldb::SBVariablesOptions &options) {
-              if (INTERRUPT_REQUESTED(dbg, 
-                    "Interrupted getting frame variables with {0} of {1} "
-                    "produced.", num_produced, num_variables))
+              if (INTERRUPT_REQUESTED(
+                      dbg,
+                      "Interrupted getting frame variables with {0} of {1} "
+                      "produced.",
+                      num_produced, num_variables))
```
revert? not related to this change?

clayborg commented on lldb/source/API/SBFrame.cpp:
```diff
@@ -947,6 +950,33 @@ SBValue SBFrame::FindRegister(const char *name) {
 }
 
+bool SBFrame::GetDescriptionWithFormat(const SBFormat &format, SBStream &output,
+                                       const char *default_value) {
```
remove default value and have clients response to returning `false`

clayborg commented on lldb/source/Target/StackFrame.cpp:
```diff
@@ -1779,18 +1779,36 @@ void StackFrame::CalculateExecutionContext(ExecutionContext &exe_ctx) {
+bool StackFrame::DumpUsingFormat(Stream &strm,
+                                 const FormatEntity::Entry *format,
+                                 llvm::StringRef frame_marker,
+                                 bool use_fallback_format_on_error) {
```
I am also leaning on having the `use_fallback_format_on_error` parameter removed and letting the user decide what to do if things fail. I am open to differing opinions here if others feel differently

bulbazord replied:
fwiw I agree with Greg here.

clayborg commented on lldb/tools/lldb-dap/JSONUtils.cpp:
```diff
@@ -785,11 +785,18 @@ llvm::json::Value CreateStackFrame(lldb::SBFrame &frame) {
+  std::string frame_name;
+  if (g_dap.frame_format.IsValid()) {
+    lldb::SBStream stream;
+    frame.GetDescriptionWithFormat(g_dap.frame_format, stream, "No value");
```
Doesn't make sense to show "No value" here, that will confuse users. we should listen to this return value and allow the standard code below to run `frame.GetDisplayFunctionName()`.

```
std::string frame_name;
lldb::SBStream stream;
if (g_dap.frame_format.IsValid() && frame.GetDescriptionWithFormat(g_dap.frame_format, stream)) 
  frame_name = stream.GetData();
else if (const char *name = frame.GetDisplayFunctionName())
  frame_name = name;

clayborg commented on lldb/tools/lldb-dap/package.json:
```diff
@@ -255,6 +255,11 @@
+							},
+							"customFrameFormat": {
+								"type": "string",
+								"description": "If non-empty, stack frames will have descriptions generated based on the provided format. See https://lldb.llvm.org/use/formatting.html for an explanation on format strings for frames. If the format is wrong, the default display names for frames will be used. This might come with a performance cost because debug information might need to be processed to generate the descri ...
```
Instead of saying "If the format is wrong", maybe say "If the format string contains errors an error string will be emitted into the debug console and the default frame names will be used".

bulbazord commented on lldb/include/lldb/API/SBFormat.h:
```diff
@@ -0,0 +1,72 @@
+  /// bool operator version of \a IsValid().
+  explicit operator bool() const;
+
+  /// \return
+  ///   \b true if and only if this object is valid and can be used for
+  ///   formatting.
+  bool IsValid() const;
```
Instead of adding `IsValid` can we just use `operator bool`? In general I'm not a fan of `IsValid` method and fallible construction.

clayborg commented on lldb/include/lldb/API/SBFormat.h:
```diff
@@ -0,0 +1,69 @@
+  /// \return
+  ///   \b true if and only if this object is valid and can be used for
+  ///   formatting.
+  explicit operator bool() const;
```
Does thius work correctly with SWIG and python? It it doesn't, then we might need to override a python operator manually to call this operator bool? Not sure if that is why we have all of the "IsValid()" stuff around

clayborg commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -824,4 +824,15 @@ bool ReplModeRequestHandler::DoExecute(lldb::SBDebugger debugger,
+  lldb::SBError error;
+  g_dap.frame_format = lldb::SBFormat::Parse(format.data(), error);
+  if (error.Fail()) {
+    llvm::errs() << "The provided frame format '" << format
+                 << "' couldn't be parsed. " << error.GetCString() << "\n";
```
Dump to the console instead of `lldb::errs()` or `llvm::outs()` as stdout/stderr go to `/dev/null` as the stdin/stdout is used for packets. When we start lldb-dap, we dup the stdin/out/err file descriptors so they don't use FD 0, 1 and 2, and then we close stdin/out/err and redirect to /dev/null. Dump to the console with:
```
    g_dap.SendOutput(OutputType::Console, llvm::StringRef(...));
```

clayborg reviewed (changes requested):
Just need to not use `llvm::errs()` or `llvm::outs()` in lldb DAP, use DAP::SendOutput(OutputType::Console, ...)` to ensure that the users sees this.

clayborg commented on lldb/source/API/SBFormat.cpp:
```diff
@@ -0,0 +1,47 @@
+using namespace lldb;
+using namespace lldb_private;
+
+SBFormat::SBFormat() : m_opaque_sp() {}
```
Do we want a constructor like:
```
Also do we want an actual constructor for SBFormat like:
```
SBFormat(const char *format, lldb::SBError &error);
```
Seems a pain to always have to write:
```
SBError error;
SBFormat format = SBFormat::Parse(<format>, error);
```
Versus:
```
SBFormat format(<format>, error);
```

walter-erquinigo replied:
good idea!

clayborg reviewed (changes requested):
One last request: test the SBFormat API in a stand alone test in `lldb/test/api/sbformat` so we can see if the errors work and the "operator bool" works as expected:
```
format = lldb.SBFormat()
self.assertFalse(format)
err = lldb.SBError()
format = lldb.SBFormat("${bad}", error)
self.assertFalse(format) # We expect an invalid object back if we have an error
self.assertTrue(error.Fail())
format = lldb.SBFormat("${frame.index}", error)
# Check the error string here as well
self.assertTrue(format)
self.assertTrue(error.Success())
```

clayborg reviewed (approved):
Thanks for all of the changes. Looks great!

walter-erquinigo committed:
Update TestFormat.py
